internal/merkle/proof_test.go

```diff
@@ -61,61 +66,110 @@ func (s *ComputeProofSuite) TestHeightOne() {
 	})
 
 	s.Run("one leaf", func() {
-		root, siblings, err := CreateProofs([]model.Hash{leaf1}, uint(height))
+		leaves := []model.Hash{leaf1}
+
+		root, siblings, err := CreateProofs(leaves, uint(height))
 		s.Require().Nil(err)
 
 		s.Equal(
 			crypto.Keccak256Hash(leaf1[:], s.pristine[:]),
 			root,
 		)
-		s.Equal(height, len(siblings))
-		s.Equal(1, len(siblings[0]))
-		s.Equal(s.pristine, siblings[0][0])
+
+		s.Equal(len(leaves)*height, len(siblings))
+		s.Equal(s.pristine, siblings[0])
+
+		leafSiblings := siblings[0 : 0+height]
+		s.Equal(root, rootFromSiblings(leaf1, 0, leafSiblings))
 	})
 
 	s.Run("two leaves", func() {
 		leaves := []model.Hash{leaf1, leaf2}
+		leavesCopy := make([]model.Hash, len(leaves))
+		copy(leavesCopy, leaves)
 
-		root, siblings, err := CreateProofs(leaves, uint(height))
+		root, siblings, err := CreateProofs(leavesCopy, uint(height))
 		s.Require().Nil(err)
 
 		s.Equal(
 			crypto.Keccak256Hash(leaf1[:], leaf2[:]),
 			root,
 		)
-		s.Equal(len(leaves), len(siblings))
-		for idx := range siblings {
-			s.Equal(height, len(siblings[idx]))
+
+		s.Equal(len(leaves)*height, len(siblings))
+		s.Equal(leaf2, siblings[0*height+0])
+		s.Equal(leaf1, siblings[1*height+0])
+
+		for idx := range leaves {
+			leafSiblings := siblings[idx*height : idx*height+height]
+			s.Equal(root, rootFromSiblings(leaves[idx], idx, leafSiblings))
 		}
-		s.Equal(leaf2, siblings[0][0])
-		s.Equal(leaf1, siblings[1][0])
 	})
 
 	s.Run("three leafs", func() {
 		leaves := make([]model.Hash, 3)
 
 		_, _, err := CreateProofs(leaves, 1)
+		s.Require().NotNil(err)
 
 		s.ErrorContains(err, "too many leaves for height")
 	})
 }
 
-func (s *ComputeProofSuite) TestHeightTwo() {
+func (s *CreateProofsSuite) TestHeightTwo() {
 	height := 2
 	leaf1 := crypto.Keccak256Hash([]byte("Merkle"))
 	leaf2 := crypto.Keccak256Hash([]byte("trees"))
 	leaf3 := crypto.Keccak256Hash([]byte("are"))
 	leaf4 := crypto.Keccak256Hash([]byte("cool"))
 
+	s.Run("no leaves", func() {
+		root, siblings, err := CreateProofs(nil, uint(height))
+		s.Require().Nil(err)
+
+		s.Equal(
+			crypto.Keccak256Hash(
+				crypto.Keccak256Hash(s.pristine[:], s.pristine[:]).Bytes(),
+				crypto.Keccak256Hash(s.pristine[:], s.pristine[:]).Bytes(),
+			),
+			root,
+		)
+
+		s.Equal(0, len(siblings))
+	})
+
+	s.Run("one leaf", func() {
+		leaves := []model.Hash{leaf1}
+
+		root, siblings, err := CreateProofs(leaves, uint(height))
+		s.Require().Nil(err)
+
+		s.Equal(
+			crypto.Keccak256Hash(
+				crypto.Keccak256Hash(leaf1[:], s.pristine[:]).Bytes(),
+				crypto.Keccak256Hash(s.pristine[:], s.pristine[:]).Bytes(),
+			),
+			root,
+		)
+
+		s.Equal(len(leaves)*height, len(siblings))
+		s.Equal(s.pristine, siblings[0])
+
+		leafSiblings := siblings[0 : 0+height]
+		s.Equal(root, rootFromSiblings(leaf1, 0, leafSiblings))
+	})
+
 	s.Run("two leaves", func() {
 		leaves := []model.Hash{leaf1, leaf2}
+		leavesCopy := make([]model.Hash, len(leaves))
+		copy(leavesCopy, leaves)
```

Comment:
So I can grab a leaf by its index later when verifying the proofs. Since `CreateProofs` uses `leaves`' underlying array to do its thing, I can't reuse it.

Reply:
I see, so we leave it up to the caller to decide whether they want to copy the array of leaves before calling `CreateProofs`.